File: 2021_12_24_advent_of_code_in_fsharp/AdventOfCode/AdventOfCode2021/Day10.fs
module AdventOfCode2021.Day10

open System

let loadData (input: string) = input.Split Environment.NewLine

type State =
    | Processing of char list
    | Corrupted of char
    | Completed of char list

let Close2OpenMap =
    Map [ (')', '(')
          ('>', '<')
          (']', '[')
          ('}', '{') ]

let Open2CloseMap = Close2OpenMap |> Map.toSeq |> Seq.map (fun (x, y) -> y, x) |> Map

let Points1 =
    Map [ (')', 3)
          (']', 57)
          ('}', 1197)
          ('>', 25137) ]

let Points2 =
    Map [ (')', 1L)
          (']', 2L)
          ('}', 3L)
          ('>', 4L) ]


let processText text =
    Seq.append text [ '-' ]
    |> Seq.scan
        (fun state c ->
            match state with
            | Processing items ->
                match c with
                | '-' -> Completed items
                | ')'
                | ']'
                | '}'
                | '>' ->
                    match items with
                    | head :: rest -> if Close2OpenMap.[c] = head then Processing rest else Corrupted c
                    | _ -> Corrupted c
                | _ -> Processing(c :: items)
            | _ -> state)
        (Processing [])

let puzzle1 (input: string) =
    let data = loadData input
    let result =
        data
        |> Seq.choose (
            processText
            >> Seq.tryPick (function
                | Corrupted c -> Some c
                | _ -> None)
        )
        |> Seq.sumBy (fun c -> Points1.[c])
    result |> string

let puzzle2 (input: string) =
    let data = loadData input
    let points =
        data
        |> Seq.choose (
            processText
            >> Seq.tryPick (function
                | Processing _ -> None
                | state -> Some state)
            >> Option.bind (function
                | Completed item -> Some item
                | _ -> None)
        )
        |> Seq.map (Seq.map (fun c -> Open2CloseMap.[c]) >> Seq.fold (fun a c -> (a * 5L) + Points2.[c]) (0L))
        |> Seq.sort
        |> Seq.toArray
    let result = points.[(points.Length - 1) / 2]
    result |> string

File: 2022_04_23_advent_of_code_in_csharp/AdventOfCode/AdventOfCode2021/Day10.cs

namespace AdventOfCode.AdventOfCode2021;

static class Day10
{
    record State;
    record Processing(LList<char>? Items) : State;
    record Corrupted(char Item) : State;
    record Completed(LList<char>? Items) : State;

    static Dictionary<char, char> Close2OpenMap = new() { { ')', '(' }, { '>', '<' }, { ']', '[' }, { '}', '{' }, };

    static Dictionary<char, char> Open2CloseMap = Close2OpenMap.ToDictionary(kv => kv.Value, kv => kv.Key);

    static Dictionary<char, int> Points1 = new() { { ')', 3 }, { ']', 57 }, { '}', 1197 }, { '>', 25137 }, };

    static Dictionary<char, long> Points2 = new() { { ')', 1L }, { ']', 2L }, { '}', 3L }, { '>', 4L }, };

    public static string[] LoadData(string input) => input.Split(Environment.NewLine);

    static IEnumerable<State> ProcessText(string text) =>
        text.Concat("-").Scan(
            new Processing(null) as State,
            (state, c) => state switch
            {
                Processing(var items) =>
                    c switch
                    {
                        '-' => new Completed(items),
                        ')' or ']' or '}' or '>' =>
                            items switch
                            {
                                (var head, var rest) when Close2OpenMap[c] == head => new Processing(rest),
                                _ => new Corrupted(c)
                            },
                        _ => new Processing(new(c, items))
                    },
                _ => state
            });

    public static string Puzzle1(string input)
    {
        var data = LoadData(input);
        var result = data
            .Choose(text => ProcessText(text).TryPick(s => s is Corrupted(var c) ? Option.Some(c) : Option.None<char>()))
            .Select(c => Points1[c])
            .Sum();
        return result.ToString();
    }

    public static string Puzzle2(string input)
    {
        var data = LoadData(input);
        var points = data
            .Choose(text => ProcessText(text)
                .TryPick(s => s is Processing ? Option.None<State>() : Option.Some(s))
                .Bind(s => s is Completed(var items) ? Option.Some(items) : Option.None<LList<char>?>()))
            .Select(items => items.ToEnumerable().Aggregate(0L, (a, c) => (a * 5) + Points2[Open2CloseMap[c]]))
            .OrderBy(x => x)
            .ToArray();
        var result = points[(points.Length - 1) / 2];
        return result.ToString();
    }
}